text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,515 @@
+
+This RFC proposes to loosen this restriction to allow custom smart pointer types to be accepted as a `self` type just like for the standard library types.
+
+See also [this blog post](https://medium.com/@adetaylor/the-case-for-stabilizing-arbitrary-self-types-b07bab22bb45), especially for a list of more specific use-cases.
+
+## Motivation for the v2 changes
+
+Unstable Rust contains an implementation of arbitrary self types based around the `Deref<Target=T>` trait. Naturally, that trait also provides a means to create a `&T`. Example:
+
+```rust
+#[feature(arbitrary_self_types)]
+
+struct SmartPtr<T>(*const T);
+
+impl<T: ?Sized> Deref for SmartPtr<T> {
+    type Target = T;
+    fn deref(&self) -> &Self::Target {
+        // never called, but smart pointers need to implement this method
+        // sometimes it's just not safe to create a reference to self.0
+    }
+}
+
+struct ConcreteType;
+
+impl ConcreteType {
+    fn some_method(self: SmartPtr<ConcreteType>) {
+
+    }
+}
+
+fn main() {
+    let concrete: SmartPtr<ConcreteType> = ...;
+    concrete.some_method();
+}
+```
+
+This works well for some smart pointer types where it's OK to create `&T` (but not necessarily `&mut T`). This includes `Pin` and the reference counted pointers. For that reason, the original arbitrary self types feature could be based around `Deref`. But in other smart pointer use-cases (especially those relating to foreign language semantics) it's not OK to create even `&T`.
+
+The arbitrary self types feature should be enhanced so it works even when we can't allow `&T`. As noted above, that's most commonly because of semantic differences to pointers in other languages, but it might be because references have special meaning or behavior in some pure Rust domain. Either way, it may not be OK to create a Rust reference `&T`, yet we may want to allow methods to be called on some reference-like thing.
+
+For this reason, implementing `Deref::deref` is problematic for many of the likely users of this "arbitrary self types" feature.
+
+If you're implementing a smart pointer `P<T>`, and you need to allow `impl T { fn method(self: P<T>) { ... }}`, yet you can't allow a reference `&T` to exist, any option for implementing `Deref::deref` has drawbacks:
+
+* Specify `Deref::Target=T` and panic in `Deref::deref`. Not good.
+* Specify `Deref::Target=*const T`. This is only possible if your smart pointer type contains a `*const T` which you can reference - this isn't the case for (for instance) weak pointers or types containing `NonNull`.
+
+Therefore, the current Arbitrary Self Types v2 provides a separate `Receiver` trait, so that there's no need to provide an awkward `Deref::deref` implementation.
+
+In addition, this v2 proposes to block generic receivers, which are currently allowed by the v1 (unstable) arbitrary self types feature. See the [diagnostics section for reasoning](#diagnostics).
```

Comment:
It might be worth elaborating on which parts of the receiver are allowed to be generic, and which parts aren't. Clearly, for diagnostic reasons, `fn<T: Receiver<Target = Self>>(self: T)` is problematic, but I'd expect that `Self` itself is permitted to be a generic type, as long as the smart pointer is definite; if not, that should probably be called out explicitly. Also, it might be a bit surprising to users if things like `fn<A: Allocator>(Box<Self, A>)` are disallowed by this rule, but I don't know whether that would run into the same issues as full generic receivers.